tokio/tests/dump.rs

```diff
@@ -147,10 +147,71 @@ mod future_completes_during_trace {
         async fn dump() {
             let handle = Handle::current();
             let _dump = handle.dump().await;
+            tokio::task::yield_now().await;
         }
 
         rt.block_on(async {
             let _ = tokio::join!(tokio::spawn(complete_during_trace()), dump());
         });
     }
 }
+
+/// Regression tests for #6051.
+///
+/// These tests ensure that tasks notified outside of a worker will not be
+/// traced, since doing so will un-set their notified bit prior to them being
+/// run and panic.
+mod notified_during_tracing {
+    use super::*;
+
+    fn test(rt: tokio::runtime::Runtime) {
+        async fn dump() {
+            loop {
+                let handle = Handle::current();
+                let _dump = handle.dump().await;
+                // Without this yield, the `current_runtime` test hangs.
+                tokio::task::yield_now().await;
+            }
+        }
+
+        rt.block_on(async {
+            let timer = tokio::spawn(async {
+                loop {
+                    tokio::time::sleep(tokio::time::Duration::from_nanos(1)).await;
+                }
+            });
+
+            let timeout = async {
+                tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            };
+
+            tokio::select!(
+                biased;
+                _ = timeout => {},
+                _ = timer => {},
+                _ = dump() => {},
+            );
+        });
+    }
+
+    #[test]
+    fn current_thread() {
+        let rt = runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        test(rt)
+    }
```

Comment:
@Darksonn This test doesn't fail before the fix — only the `multi_thread` test fails. I've included the `current_thread` variant for completeness, but is #6051 even an issue that can occur on the `current_thread` runtime?

Reply:
It probably can if you spawn from outside the runtime. But `block_on` isn't outside the runtime in the same way with the current-thread runtime.

Reply:
Hm, something like this?
```rust
#[test]
fn current_thread() {
    let rt = runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    let other_rt = runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    async fn dump() {
        loop {
            let handle = Handle::current();
            let _dump = handle.dump().await;
            tokio::task::yield_now().await;
        }
    }

    // spawn the timer on `other_rt`
    let timer = other_rt.spawn(async {
        loop {
            tokio::time::sleep(tokio::time::Duration::from_nanos(1)).await;
        }
    });

    rt.block_on(async {
        let timeout = async {
            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
        };

        tokio::select!(
            biased;
            _ = timeout => {},
            _ = timer => {},
            _ = dump() => {},
        );
    });
}
```
(This doesn't yet repro the issue, but I'm curious if I'm going in the right direction.)